File: Problem/BOJ_11909_배열 탈출/main.py
from sys import stdin
from heapq import heappush, heappop

moving_dir = [[0, 1], [1, 0]]

stdin = open("./input.txt", "r")
n = int(stdin.readline())

array = [[0] * (n + 1)]
for _ in range(n):
    array.append([0] + list(map(int, stdin.readline().split())))
dist = [[float("INF")] * (n + 1) for _ in range(n + 1)]
visited = [[False] * (n + 1) for _ in range(n + 1)]


def dijkstra():
    heap = []
    heappush(heap, (0, 1, 1))
    dist[1][1] = 0
    visited[1][1] = True

    while heap:
        cur_dist, cur_row, cur_col = heappop(heap)

        if cur_dist > dist[cur_row][cur_col]:
            continue

        for moving_row, moving_col in moving_dir:
            next_row = cur_row + moving_row
            next_col = cur_col + moving_col
            if 1 <= next_row <= n and 1 <= next_col <= n:
                if not visited[next_row][next_col]:
                    if array[cur_row][cur_col] > array[next_row][next_col] and cur_dist < dist[next_row][next_col]:
                        dist[next_row][next_col] = cur_dist
                        heappush(heap, (dist[next_row][next_col], next_row, next_col))
                    else:
                        count = array[next_row][next_col] - array[cur_row][cur_col] + 1
                        if cur_dist + count < dist[next_row][next_col]:
                            dist[next_row][next_col] = cur_dist + count
                            heappush(heap, (dist[next_row][next_col], next_row, next_col))
                    visited[next_row][next_col] = True
    return dist


def main():
    dijkstra()
    print(dist[n][n])


if __name__ == '__main__':
    main()

File: Problem/BOJ_11909_배열 탈출/src/Main.java
import java.util.*;
import java.io.*;


public class Main {
    static int n;
    static int [][] movingDir = {{0, 1}, {1, 0}};
    static int[][] array;
    static int[][] dist;
    static StringTokenizer st;

    public static void dijkstra() {
        PriorityQueue<PositionInfo> pq = new PriorityQueue<>();
        pq.add(new PositionInfo(0, 1, 1));
        dist[1][1] = 0;

        while (pq.size() != 0) {
            PositionInfo curInfo = pq.poll();
            int curDist = curInfo.dist;
            int curRow = curInfo.row;
            int curCol = curInfo.col;

            if (curDist > dist[curRow][curCol]) {
                continue;
            }
            for (int i = 0; i < 2; i++) {
                int movingRow = movingDir[i][0];
                int movingCol = movingDir[i][1];
                int nextRow = curRow + movingRow;
                int nextCol = curCol + movingCol;
                if (1 <= nextRow && nextRow <= n && 1 <= nextCol && nextCol <= n) {
                    if (array[curRow][curCol] > array[nextRow][nextCol]) {
                        if (curDist < dist[nextRow][nextCol]) {
                            dist[nextRow][nextCol] = curDist;
                            pq.add(new PositionInfo(dist[nextRow][nextCol], nextRow, nextCol));
                        }
                    } else {
                        int count = array[nextRow][nextCol] - array[curRow][curCol] + 1;
                        if (curDist + count < dist[nextRow][nextCol]) {
                            dist[nextRow][nextCol] = curDist + count;
                            pq.add(new PositionInfo(dist[nextRow][nextCol], nextRow, nextCol));
                        }
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws IOException{
        System.setIn(new FileInputStream("src\\input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        array = new int[n + 1][n + 1];
        dist = new int[n + 1][n + 1];

        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= n; j++) {
                array[i][j] = Integer.parseInt(st.nextToken());
                dist[i][j] = Integer.MAX_VALUE;
            }
        }

        dijkstra();
        System.out.println(dist[n][n]);

//        for (int i = 0; i <= n; i++) {
//            System.out.println(Arrays.toString(dist[i]));
//        }
    }
}

class PositionInfo implements Comparable<PositionInfo> {
    public int dist;
    public int row;
    public int col;

    PositionInfo(int dist, int row, int col) {
        this.dist = dist;
        this.row = row;
        this.col = col;
    }

    @Override
    public String toString() {
        return "PositionInfo{" +
                "dist=" + dist +
                ", row=" + row +
                ", col=" + col +
                '}';
    }

    @Override
    public int compareTo(PositionInfo positionInfo) {
        return this.dist - positionInfo.dist;
    }
}